cmd/promtool/sd.go

```diff
@@ -0,0 +1,165 @@
+	logger := log.NewLogfmtLogger(log.NewSyncWriter(os.Stderr))
+
+	cfg, err := config.LoadFile(sdConfigFiles, false, logger)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, "Cannot load config", err)
+		return 2
+	}
+
+	var discoveryConfigs discovery.Configs
+	var relabelConfigs []*relabel.Config
+	jobs := []string{}
+	jobMatched := false
+	for _, v := range cfg.ScrapeConfigs {
+		jobs = append(jobs, v.JobName)
+		if v.JobName == sdJobName {
+			jobMatched = true
+			discoveryConfigs = v.ServiceDiscoveryConfigs
+			relabelConfigs = v.RelabelConfigs
+			break
+		}
+	}
+
+	if !jobMatched {
+		fmt.Fprintf(os.Stderr, "Job %s not found. Select one of:\n", sdJobName)
+		for _, job := range jobs {
+			fmt.Fprintf(os.Stderr, "\t%s\n", job)
+		}
+		return 1
+	}
+
+	targetGroupChan := make(chan []*targetgroup.Group)
+	ctx, cancel := context.WithTimeout(context.Background(), sdTimeout)
+	defer cancel()
+
+	for _, cfg := range discoveryConfigs {
+		d, err := cfg.NewDiscoverer(discovery.DiscovererOptions{Logger: logger})
+		if err != nil {
+			fmt.Fprintln(os.Stderr, "Could not create new discoverer", err)
+			return 2
+		}
+		go d.Run(ctx, targetGroupChan)
+	}
+
+	var targetGroups []*targetgroup.Group
+	sdCheckResults := make(map[string][]*targetgroup.Group)
+	returnCode := 0
+	loop := true
+	for loop {
+		select {
+		case targetGroups = <-targetGroupChan:
+			for _, tg := range targetGroups {
+				sdCheckResults[tg.Source] = append(sdCheckResults[tg.Source], tg)
+			}
+		case <-ctx.Done():
+			loop = false
+			break
+		}
+	}
+
+	i := 0
+	fmt.Println("[")
+	for _, tgs := range sdCheckResults {
+		for _, result := range getSDCheckResult(tgs, relabelConfigs) {
+			res, err := json.MarshalIndent(result, "  ", "  ")
+			if err != nil {
+				fmt.Fprintf(os.Stderr, `{"error": "Could not marshal result json: %s"}\n`, err)
+				returnCode = 2
+			}
+
+			if i > 0 {
+				fmt.Println(",")
+			}
+			fmt.Print("  ", string(res))
+			i++
+		}
+	}
+	if i > 0 {
+		fmt.Println("")
+	}
+	fmt.Println("]")
+	return returnCode
+}
+
+func getSDCheckResult(targetGroups []*targetgroup.Group, relabelConfigs []*relabel.Config) []sdCheckResult {
+	sdCheckResults := []sdCheckResult{}
+	for _, targetGroup := range targetGroups {
+		for _, target := range targetGroup.Targets {
+			labelSlice := make([]labels.Label, 0, len(target)+len(targetGroup.Labels))
+
+			var source string
+			for name, value := range target {
+				if name == "__address__" {
+					source = string(value)
+				}
+				labelSlice = append(labelSlice, labels.Label{Name: string(name), Value: string(value)})
+			}
+
+			for name, value := range targetGroup.Labels {
+				if _, ok := target[name]; !ok {
+					labelSlice = append(labelSlice, labels.Label{Name: string(name), Value: string(value)})
+				}
+			}
+
+			targetLabels := labels.New(labelSlice...)
+			result := sdCheckResult{
+				Target:          source,
+				OriginalLabels:  targetLabels,
+				AfterRelabeling: relabel.Process(targetLabels, relabelConfigs...),
```

Comment:
We should remove the labels that starts with __, except the reserved labels. Just like Prometheus does.

Could we just export and call `populateLabels` ?